src/services/document-service.js

```diff
@@ -90,9 +89,28 @@ class DocumentService {
     return this.documentStore.find({ _id: { $in: documentKeys } }, { sort: lastUpdatedFirst, projection: metadataProjection });
   }
 
-  async getDocumentsByTags(searchTags) {
-    const result = await this.documentStore.find({ tags: { $all: searchTags } }, { projection: searchResultsProjection });
-    return result || [];
+  async getDocumentsByTags(searchQuery) {
+    const searchTags = new Set(searchQuery.trim()
+      .split(/\s+/)
+      .map(tag => tag.toLowerCase())
+      .filter(tag => (/^\w{3,30}$/).test(tag)));
+
+    const query = {
+      $or: Array.from(searchTags).map(tag => ({
+        tags: {
+          $regex: `.*${tag}.*`, $options: 'i'
+        }
+      }))
+    };
+
+    const documents = await this.documentStore
+      .find(query, { projection: searchResultsProjection }) || [];
+
+    return documents.map(result => ({
+      ...result,
+      tagMatchCount: result.tags
+        .reduce((count, tag) => searchTags.has(tag.toLowerCase()) ? count + 1 : count, 0)
```

Comment:
I think this is a bit overcomplicated, could just be
```
result.tags.filter(tag => searchTags.has(tag.toLowerCase())).length;
```

Reply:
A reduce is semantically more correct, but okay.

Reply:
Don't see it so, you want a subset that fits a requirement 🤷 You don't sum up/gather up info from each iterated item.

Reply:
I don't want the subset, I want one number :). I am reducing a collection to a number.

Reply:
Yeah, by filtering and counting how many you get 😄

Reply:
What I'm trying to say is that I don't need the filtered instances, I only need a number. You can get that number in either of the 2 ways, somehow to me a reduce makes more sense since it's the intent here. Anyway, the code now uses filter :).

Reply:
Can we all agree that `map` and `filter` are both subsets of `reduce` that are easier to write and read ? :laughing:

Reply:
Definitely, which makes `reduce` overcomplicated at times 😁